casper/src/test/scala/coop/rchain/casper/util/rholang/MergingBranchMergerSpec.scala

```diff
@@ -0,0 +1,469 @@
+    dirs <- Resources.copyStorage[Task](genesisContext.storageDirectory)
+    rm   <- Resources.mkRuntimeManagerAt[Task](dirs.rspaceDir)()
+  } yield rm
+
+//  private def txTerm(from: String = "", to: String = "") =
+//    Source
+//      .fromResource("RevTx.rho")
+//      .mkString
+//      .replaceAll("%from%", from)
+//      .replaceAll("%to%", to)
+//      .replaceAll("%amount%", 100000.toString)
+
+  private def biTxTerm(from: String, to: String) =
+    Source
+      .fromResource("RevTxBackAndForth.rho")
+      .mkString
+      .replaceAll("%from%", from)
+      .replaceAll("%to%", to)
+      .replaceAll("%amt%", 1000000.toString)
+      .replaceAll("%amt_back%", 100.toString)
+
+  def makeTxAndCommitState(
+      runtimeManager: RuntimeManager[Task],
+      baseState: StateHash,
+      payerKey: PrivateKey,
+      validator: PublicKey,
+      seqNum: Int = 0,
+      blockNum: Long = 0
+  ): Task[(StateHash, Seq[ProcessedDeploy], Seq[ProcessedSystemDeploy])] =
+    for {
+      txDeploy    <- ConstructDeploy.sourceDeployNowF("Nil", sec = payerKey)
+      userDeploys = txDeploy :: Nil
+      systemDeploys = CloseBlockDeploy(
+        SystemDeployUtil
+          .generateCloseDeployRandomSeed(
+            validator,
+            seqNum
+          )
+      ) :: Nil
+      blockData = BlockData(
+        txDeploy.data.timestamp,
+        blockNum,
+        validator,
+        seqNum
+      )
+      invalidBlocks = Map.empty[BlockHash, Validator]
+      r <- runtimeManager.computeState(baseState)(
+            userDeploys,
+            systemDeploys,
+            blockData,
+            invalidBlocks
+          )
+    } yield r
+
+  def mkStatesSteams(preStateHash: StateHash, seqNum: Int, blockNum: Long)(
+      implicit runtimeManager: RuntimeManager[Task]
+  ) =
+    (genesisContext.validatorPks zip genesisContext.genesisVaultsSks).map {
+      case (validatorPk, payerSk) =>
+        Stream
+          .eval(
+            makeTxAndCommitState(
+              runtimeManager,
+              preStateHash,
+              payerSk,
+              validatorPk,
+              seqNum,
+              blockNum
+            )
+          )
+    }.toList
+
+  def mkState(stateHash: StateHash, seqNum: Int, blockNum: Long, n: Int)(
+      implicit runtimeManager: RuntimeManager[Task]
+  ) =
+    mkStatesSteams(stateHash, seqNum, blockNum).get(n.toLong).get.compile.lastOrError
+
+  def mkHeadState(stateHash: StateHash, seqNum: Int, blockNum: Long)(
+      implicit runtimeManager: RuntimeManager[Task]
+  ) =
+    mkStatesSteams(stateHash, seqNum, blockNum).head.compile.lastOrError
+
+  def mkTailStates(stateHash: StateHash, seqNum: Int, blockNum: Long)(
+      implicit runtimeManager: RuntimeManager[Task]
+  ) =
+    Stream
+      .emits(mkStatesSteams(stateHash, seqNum, blockNum).tail)
+      .parJoinUnbounded
+      .compile
+      .toList
+
+  "serial execution and merging" should "be equal" in effectTest {
```

Comment:
Just want to say this is a very nice test with wisdom.

Reply:
kudos to @tgrospic